### PR TITLE
Earn: Convert controller to typescript

### PR DESCRIPTION
**Pull request by edanzer**
## Proposed Changes

**You can see the specific changes in this commit**: https://github.com/Automattic/wp-calypso/pull/81465/commits/bb80e53d10927f25462582c24159730b9222d0fc

Refactors the Earn controller to TypeScript. Because of the file name change from .js to .ts, Github is not seeing these as the same file. That can make it harder to review specific changes. So I've pushed the commits such that the file rename and the actual changes are separate, and the specific changes are visible in the second commit.

## Testing Instructions

Checkout this branch, go to `http://calypso.localhost:3000/earn/YOURDOMAIN`, and click around different tabs and sections to confirm navigation still works the same as before. You'll want to connect Stripe if you haven't so that you can navigate between sections behind the Connect to Stripe screen.

### CODE REVIEW

File: client/my-sites/earn/controller.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import page from 'page';
+import { createElement, FC } from 'react';
+import Main from './main';
+import type { Context as PageJSContext } from 'page';
+
```
Comment (edanzer) [inline client/my-sites/earn/controller.ts:5]: Added imports for FC from React and for PageJSContext.

File: client/my-sites/earn/controller.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		page.redirect( '/earn/ads-earnings/' + context.params.site_id );
+	},
+	redirectToAdsSettings: function ( context: PageJSContext ) {
+		page.redirect( '/earn/ads-settings/' + context.params.site_id );
+	},
+	layout: function ( context: PageJSContext, next: () => void ) {
+		// Scroll to the top
+		if ( typeof window !== 'undefined' ) {
+			window.scrollTo( 0, 0 );
+		}
```
Comment (edanzer) [inline client/my-sites/earn/controller.ts:17]: For all methods above, change is to type the `context` param as PageJSContext, and to type the `next` param as () => void.

File: client/my-sites/earn/controller.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+import page from 'page';
+import { createElement, FC } from 'react';
+import Main from './main';
+import type { Context as PageJSContext } from 'page';
```
Comment (edanzer) [inline client/my-sites/earn/controller.ts:4]: For imports, changes are to add imports for FC from React and for PageJSContext.

File: client/my-sites/earn/controller.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+		context.primary = createElement(
+			Main as FC< { section: string; path: string; query: string } >,
+			{
+				section: context.params.section,
+				path: context.path,
+				query: context.query,
+			}
+		);
+		next();
+	},
```
Comment (edanzer) [inline client/my-sites/earn/controller.ts:28]: For lines 19-28, change was to type Main in the createElement call as `FC< { section: string; path: string; query: string } >`.

**Review by lezama — APPROVED** [review]

**Pull request merged**